src/linear_algebra/invariant_submodule.lean

```diff
@@ -0,0 +1,142 @@
+  exact h (submodule.coe_mem _),
+end
+
+lemma invariant_under_of_proj_comp_self_comp_proj_eq
+  (h : ∀ x : E, ↑(pᵤ (T ↑(pᵤ x))) = T ↑(pᵤ x)) : U.invariant_under T :=
+begin
+  intros u hu,
+  rw [submodule.mem_comap, ← submodule.linear_proj_of_is_compl_eq_self_iff hUV _,
+      ← (submodule.linear_proj_of_is_compl_eq_self_iff hUV u).mpr hu, h],
+end
+
+lemma proj_comp_self_comp_proj_eq_iff_invariant_under :
+  U.invariant_under T ↔ (∀ x : E, ↑(pᵤ (T ↑(pᵤ x))) = T ↑(pᵤ x)) :=
+⟨U.proj_comp_self_comp_proj_eq_of_invariant_under V hUV T,
+ U.invariant_under_of_proj_comp_self_comp_proj_eq V hUV T⟩
+
+lemma proj_comp_self_comp_proj_eq_iff_invariant_under' :
+  V.invariant_under T ↔ (∀ x : E, (pᵤ (T ↑(pᵤ x)) : E) = pᵤ (T x)) :=
+by simp_rw [submodule.proj_comp_self_comp_proj_eq_iff_invariant_under _ _ hUV.symm,
+            linear_proj_of_is_compl_eq_self_sub_linear_proj, map_sub, sub_eq_self,
+            submodule.coe_sub, sub_eq_zero, eq_comm]
+
+lemma compl_invariant_under_iff_linear_proj_and_T_commute :
+  (U.invariant_under T ∧ V.invariant_under T) ↔ commute (U.subtype.comp (pᵤ)) T :=
+begin
+  simp_rw [commute, semiconj_by, linear_map.ext_iff, linear_map.mul_apply,
+           linear_map.comp_apply, U.subtype_apply],
+  split,
+  { rintros ⟨h1, h2⟩ x,
+    rw [← (U.proj_comp_self_comp_proj_eq_iff_invariant_under' V _ _).mp h2 x],
+    exact (submodule.linear_proj_of_is_compl_eq_self_iff hUV _).mpr
+      (h1 (set_like.coe_mem (pᵤ x))) },
+  { intros h,
+    split,
+    { simp_rw [U.proj_comp_self_comp_proj_eq_iff_invariant_under _ hUV, h,
+               submodule.linear_proj_of_is_compl_idempotent hUV],
+      exact λ x, rfl },
+    { simp_rw [U.proj_comp_self_comp_proj_eq_iff_invariant_under' _ hUV, h,
+               submodule.linear_proj_of_is_compl_idempotent hUV],
+      exact λ x, rfl } }
+end
+
+lemma commutes_with_linear_proj_iff_linear_proj_eq [invertible T] :
+  commute (U.subtype.comp pᵤ) T ↔
+    (⅟ T).comp ((U.subtype.comp pᵤ).comp T) = U.subtype.comp pᵤ :=
+begin
+  rw [commute, semiconj_by],
+  simp_rw [linear_map.mul_eq_comp],
+  split,
+  { intro h,
+    simp_rw [h, ← linear_map.mul_eq_comp, ← mul_assoc, inv_of_mul_self, one_mul], },
+  { intros h,
+    rw ← h, simp_rw [← linear_map.mul_eq_comp, ← mul_assoc, mul_inv_of_self],
+    rw [mul_assoc (⅟ T) _ _],
+    simp_rw [linear_map.mul_eq_comp, h], refl, }
+end
+
+lemma invariant_under_inv_iff_U_subset_image [invertible T] :
+  U.invariant_under (⅟ T) ↔ ↑U ⊆ T '' U :=
+begin
+  split,
+ { intros h x hx,
+   simp only [set.mem_image, set_like.mem_coe],
+   use (⅟ T) x,
+   rw [← linear_map.comp_apply, ← linear_map.mul_eq_comp,
+       mul_inv_of_self, linear_map.one_apply, eq_self_iff_true, and_true],
+   exact h hx, },
+ { intros h x hx,
+   rw submodule.mem_comap,
+   simp only [set.subset_def, set.mem_image] at h,
+   cases h x hx with y hy,
+   rw [← hy.2, ← linear_map.comp_apply, ← linear_map.mul_eq_comp,
+       inv_of_mul_self],
+   exact hy.1 }
+end
```

Comment:
If you want to go the `equiv` road, you can do
```suggestion
lemma invariant_under_inv_iff_U_subset_image {S : E ≃ₗ[R] E} :
  U.invariant_under S.symm ↔ ↑U ⊆ S '' U :=
(U.invariant_under_iff S.symm).trans (S.to_equiv.symm.subset_image' _ _).symm
```

(And if you want to use `invertible`, you should probably find a way to use `equiv.image_eq_preimage` or a variant)

Reply:
I shortened the proofs as suggested (by finding a way to use `equiv.image_eq_preimage`) and kept `invertible`